### ISSUE TITLE
NAN in _fit_global_rotations()

### ISSUE THREAD

**Issue by siddharth1712** [isarandi/nlf#28]
Hi,

I am trying to add your SMPL fitter as a frozen pipeline into my larger trainable model to get the SMPL shape parameters from an image tensor. The fitting works well but during back-propagation through your pipeline, the gradients are becoming Nan. I narrowed this down and found that during forward pass when the function _fit_global_rotations() is called  it produces Nans for some joints, namely joint 8,19,21,23,24 and all joints from 40 to 54. I am using SMPL-X. 
Can you help in telling why is this the case, or how can i overcome it?
Below is the code of _fit_global_rotation along with added print statements for debugging

```
def _fit_global_rotations(
        self,
        target_vertices: torch.Tensor,
        target_joints: Optional[torch.Tensor],
        reference_vertices: torch.Tensor,
        reference_joints: Optional[torch.Tensor],
        vertex_weights: Optional[torch.Tensor],
        joint_weights: Optional[torch.Tensor],
    ) -> torch.Tensor:
        device = target_vertices.device
        glob_rots = []
        mesh_weight = torch.tensor(1e-6, device=device, dtype=torch.float32)
        joint_weight = 1 - mesh_weight

        if target_joints is None or reference_joints is None:
            target_joints = self.body_model.J_regressor_post_lbs @ target_vertices
            reference_joints = self.body_model.J_regressor_post_lbs @ reference_vertices

        part_assignment = torch.argmax(self.body_model.weights, dim=1)
        # Disable the rotation of toes separately from the feet
        part_assignment = torch.where(
            part_assignment == 10,
            torch.tensor(7, dtype=torch.int64, device=device),
            part_assignment,
        )
        part_assignment = torch.where(
            part_assignment == 11,
            torch.tensor(8, dtype=torch.int64, device=device),
            part_assignment,
        )

        for i in range(self.body_model.num_joints):
            # Disable the rotation of toes separately from the feet
            if i == 10:
                glob_rots.append(glob_rots[7])
                continue
            elif i == 11:
                glob_rots.append(glob_rots[8])
                continue

            selector = torch.where(part_assignment == i)[0]
            default_body_part = reference_vertices[:, selector]
            estim_body_part = target_vertices[:, selector]
            weights_body_part = (
                vertex_weights[:, selector].unsqueeze(-1) * mesh_weight
                if vertex_weights is not None
                else mesh_weight
            )

            default_joints = reference_joints[:, self.children_and_self[i]]
            estim_joints = target_joints[:, self.children_and_self[i]]
            weights_joints = (
                joint_weights[:, self.children_and_self[i]].unsqueeze(-1) * joint_weight
                if joint_weights is not None
                else joint_weight
            )

            body_part_mean_reference = torch.mean(default_joints, dim=1, keepdim=True)
            default_points = torch.cat(
                [
                    (default_body_part - body_part_mean_reference) * weights_body_part,
                    (default_joints - body_part_mean_reference) * weights_joints,
                ],
                dim=1,
            )

            body_part_mean_target = torch.mean(estim_joints, dim=1, keepdim=True)
            estim_points = torch.cat(
                [
                    (estim_body_part - body_part_mean_target),
                    (estim_joints - body_part_mean_target),
                ],
                dim=1,
            )
            
            print(f"--- Joint {i} ---")
            print(f"  selector size: {selector.shape}")
            print(f"  default_body_part mean: {default_body_part.mean().item():.4f}, std: {default_body_part.std().item():.4f}")
            print(f"  estim_body_part mean: {estim_body_part.mean().item():.4f}, std: {estim_body_part.std().item():.4f}")
            print(f"  default_joints range: {default_joints.min().item():.4f} to {default_joints.max().item():.4f}")
            print(f"  estim_joints range: {estim_joints.min().item():.4f} to {estim_joints.max().item():.4f}")
            print(f"  body_part_mean_reference norm: {body_part_mean_reference.norm().item():.4f}")
            if weights_body_part.numel() > 0:
                print(f"  weights_body_part min: {weights_body_part.min().item():.6f}, max: {weights_body_part.max().item():.6f}")
            else:
                print(f"  ⚠️ weights_body_part is empty for joint {i}")

            
            glob_rot = kabsch(estim_points, default_points)
            glob_rots.append(glob_rot)
```

The output looks like below

--- Joint 21 ---
  selector size: torch.Size([0])
  default_body_part mean: nan, std: nan
  estim_body_part mean: nan, std: nan
  default_joints range: -0.3283 to 0.3937
  estim_joints range: -0.3282 to 0.3419
  body_part_mean_reference norm: 0.4621
  ⚠️ weights_body_part is empty for joint 21
--- Joint 22 ---
  selector size: torch.Size([23])
  default_body_part mean: -0.1500, std: 0.1012
  estim_body_part mean: -0.1510, std: 0.0986
  default_joints range: -0.3545 to -0.0075
  estim_joints range: -0.3552 to -0.0101
  body_part_mean_reference norm: 0.3619
  weights_body_part min: 0.000001, max: 0.000002
--- Joint 23 ---
  selector size: torch.Size([0])
  default_body_part mean: nan, std: nan
  estim_body_part mean: nan, std: nan
  default_joints range: -0.3640 to -0.0297
  estim_joints range: -0.3645 to -0.0320
  body_part_mean_reference norm: 0.3778
  ⚠️ weights_body_part is empty for joint 23
--- Joint 24 ---
  selector size: torch.Size([0])
  default_body_part mean: nan, std: nan
  estim_body_part mean: nan, std: nan
  default_joints range: -0.3651 to -0.0968
  estim_joints range: -0.3655 to -0.0976
  body_part_mean_reference norm: 0.3905
  ⚠️ weights_body_part is empty for joint 24
--- Joint 25 ---
  selector size: torch.Size([3])
  default_body_part mean: 0.1337, std: 0.1773
  estim_body_part mean: 0.1353, std: 0.1481
  default_joints range: -0.0448 to 0.3757
  estim_joints range: -0.0201 to 0.3300
  body_part_mean_reference norm: 0.3737
  weights_body_part min: 0.000000, max: 0.000000

**Comment by siddharth1712** [comment]
Hi,

I was able to avoid the above error by using SMPL instad of SMPLX, but now face a new error where while doing back propagation through the 
```
    return torch.nn.functional.grid_sample(
        image.unsqueeze(0), old_coords_normalized.unsqueeze(0),
        align_corners=True, mode='bilinear', padding_mode='zeros').squeeze(0)
```
in the warp_single_image() function in warping.py, I am observing Nan in the gradients. I then saw that the old_coords_normalized were exceeding the range [-1,1] but even after clamping them [-1,1] range I am getting Nan in the gradients thrugh this operation. I have checked my inputs and there are no Nan or inf in them.

**Comment by siddharth1712** [comment]
Figured it out, closing the issue
